rust/src/tx_builder.rs

```diff
@@ -154,13 +154,94 @@ pub enum CoinSelectionStrategyCIP2 {
 
 #[wasm_bindgen]
 #[derive(Clone, Debug)]
-pub struct TransactionBuilder {
-    coins_per_utxo_word: BigNum,
-    pool_deposit: BigNum,
-    key_deposit: BigNum,
-    max_value_size: u32,
-    max_tx_size: u32,
+pub struct TransactionBuilderConfig {
     fee_algo: fees::LinearFee,
+    pool_deposit: BigNum,      // protocol parameter
+    key_deposit: BigNum,       // protocol parameter
+    max_value_size: u32,       // protocol parameter
+    max_tx_size: u32,          // protocol parameter
+    coins_per_utxo_word: Coin, // protocol parameter
+    prefer_pure_change: bool,
+}
+
+#[wasm_bindgen]
+#[derive(Clone, Debug)]
+pub struct TransactionBuilderConfigBuilder {
+    fee_algo: Option<fees::LinearFee>,
```

Comment:
It might also be worth thinking about that since most of these are protocol parameters maybe it would make sense to treat the things that are protocol parameters (and thus will only have 1 valid setting that must be provided and won't change within a program and could be stored somewhere to be reused) differently from things that could depend based on use-case within a single program?

Reply:
We can look into, but overall it kinda sounds like a global static config to me so far, which is known to cause nothing but problems. Makes sense for a frontend app which is running in some context, but not for a library. E.g. if it would be something like:

```
const networkContext1 = RustModule.NetworkContextBuilder.new()
   .fee_algo(linearFee)
   .key_deposit(keyDeposit)
   ...
   .coins_per_utxo_word(coinsPerUtxoWord)
   .build()

const txBuilder1 = networkContext1.newTxBuilder();
```

Then this might work cuz we would just create a place where it loads that network context and then reuses it from all places. Even tho it's not much different from having a place that would similarly just load the tx-builder in a similar way.

But if you mean something like:
```
RustModule.set_fee_algo(linearFee);

RustModule.TransactionBuilder.new(); // knows the fee algo automatically from the global value
```

Then this would be a big no-no that causes nothing but pain and suffering long-term )

E.g. we should be able to create tx-builders for different network setups and different configurations even when running from a single frontend app instance, for example to support both testnet and mainet, or various additional networks, etc.

Reply:
@vsubhuman 

>But if you mean something like:

No, I meant more something like a separate `ProtocolParemters`-like struct to set those things then your builder config would like:

```rust
var protocol_params = ProtocolParams::new().coins_per_utxo_word(100).fee_algo(linearFee)...build();
// dev can now have 1 place to get their protocol params even if their non-protocol-param parameters to the tx builder vary
builder_config.protocol_params(&protocol_params).prefer_pure_change(true).build().unwrap();
```

So that we would separate more the protocol parameters (which at a given point in time on a chain should be the same) from the things specific just to the tx builder.

Although it depends more on how many other non-protocol-params we are adding to the builder. I guess you could have the built so far config builder with the params set saved somewhere then after #270 it copies instead of consumes so you just use that mostly-filled builder to create your specific-usage config so maybe it's not as important now that it's non-consuming.

Reply:
Ah, understood now. Yeah, I think so far there's only one non-protocol parameter so far, but this might be a really good idea for the future development if that functionality will be growing and growing further